From 820a35935045cc3fe0e40ae42de30787790e9035 Mon Sep 17 00:00:00 2001
From: Robert Iannucci <iannucci@chromium.org>
Date: Fri, 16 Feb 2024 13:49:48 -0800
Subject: [PATCH] Make autoconf relocatable.

This is the package definition for the autoconf tools and includes a patch to
make them relocatable (written against autoconf 2.69).

By default autoconf hard-codes the --prefix value into the binaries it deploys,
making them unsuitable for relocatable deployment (e.g. with CIPD). The patch
here replaces all the hard-coded paths with either:
  * The assumption that the tool is in $PATH, and so "/path/to/tool" is replaced
    by "tool"
  * The assumption that the data files are relative to the binary being run,
    e.g. if we're running ".../bin/tool" that we can find the data files at
    ".../share/extra_files".

The patch was made by doing `make install` on the base package, looking for
absolute paths and then changing the sources so the absolute paths no longer
showed up in the output of `make install`.
---
 bin/autoconf.as   |  6 ++++--
 bin/autoheader.in |  6 ++++--
 bin/autom4te.in   | 11 ++++++++---
 bin/autoreconf.in | 10 ++++++----
 bin/autoscan.in   |  8 +++++---
 bin/autoupdate.in | 12 +++++++++---
 bin/ifnames.in    |  4 +++-
 lib/autom4te.in   |  4 ----
 man/autoconf.1    |  4 ++--
 man/autoheader.1  |  4 ++--
 man/autom4te.1    |  4 ++--
 man/autoreconf.1  |  4 ++--
 man/autoscan.1    |  4 ++--
 man/autoupdate.1  |  4 ++--
 man/ifnames.1     |  4 ++--
 15 files changed, 53 insertions(+), 36 deletions(-)

diff --git a/bin/autoconf.as b/bin/autoconf.as
index 1407739..007892f 100644
--- a/bin/autoconf.as
+++ b/bin/autoconf.as
@@ -89,8 +89,10 @@ exit_missing_arg='
 # restore font-lock: '
 
 # Variables.
-: ${AUTOM4TE='@bindir@/@autom4te-name@'}
-: ${trailer_m4='@pkgdatadir@/autoconf/trailer.m4'}
+: ${AUTOM4TE='@autom4te-name@'}
+: ${PREFIX_BIN=`dirname "$0"`}
+: ${PREFIX=`dirname $PREFIX_BIN`}
+: ${trailer_m4="$PREFIX/share/autoconf/autoconf/trailer.m4"}
 autom4te_options=
 outfile=
 verbose=false
diff --git a/bin/autoheader.in b/bin/autoheader.in
index 1cbf509..baff717 100644
--- a/bin/autoheader.in
+++ b/bin/autoheader.in
@@ -32,7 +32,9 @@ use warnings FATAL => 'all';
 
 BEGIN
 {
-  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || '@pkgdatadir@';
+  use File::Basename;
+  my $PREFIX = dirname(dirname(__FILE__));
+  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || "$PREFIX/share/autoconf";
   unshift @INC, "$pkgdatadir";
 
   # Override SHELL.  On DJGPP SHELL may not be set to a shell
@@ -54,7 +56,7 @@ use Autom4te::XFile;
 our ($config_h, %symbol, %verbatim);
 
 # Lib files.
-my $autom4te = $ENV{'AUTOM4TE'} || '@bindir@/@autom4te-name@';
+my $autom4te = $ENV{'AUTOM4TE'} || '@autom4te-name@';
 my $config_h_in;
 my @prepend_include;
 my @include;
diff --git a/bin/autom4te.in b/bin/autom4te.in
index 7ebe419..e6b3875 100644
--- a/bin/autom4te.in
+++ b/bin/autom4te.in
@@ -28,7 +28,9 @@ use warnings FATAL => 'all';
 
 BEGIN
 {
-  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || '@pkgdatadir@';
+  use File::Basename;
+  my $PREFIX = dirname(dirname(__FILE__));
+  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || "$PREFIX/share/autoconf";
   unshift @INC, $pkgdatadir;
 
   # Override SHELL.  On DJGPP SHELL may not be set to a shell
@@ -47,8 +49,11 @@ use Autom4te::FileUtils;
 use Autom4te::General;
 use Autom4te::XFile;
 
+my $PREFIX = dirname(dirname(__FILE__));
+my $SHARE_DIR = "$PREFIX/share/autoconf";
+
 # Data directory.
-my $pkgdatadir = $ENV{'AC_MACRODIR'} || '@pkgdatadir@';
+my $pkgdatadir = $ENV{'AC_MACRODIR'} || $SHARE_DIR;
 
 # $LANGUAGE{LANGUAGE} -- Automatic options for LANGUAGE.
 my %language;
@@ -335,7 +340,7 @@ sub parse_args ()
 
   # Process the arguments for real this time.
   my @trace;
-  my @prepend_include;
+  my @prepend_include = ($SHARE_DIR);
   my @warnings;
 
   getopt
diff --git a/bin/autoreconf.in b/bin/autoreconf.in
index ec391a6..2d57329 100644
--- a/bin/autoreconf.in
+++ b/bin/autoreconf.in
@@ -32,7 +32,9 @@ use warnings FATAL => 'all';
 my $buildauxdir;
 BEGIN
 {
-  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || '@pkgdatadir@';
+  use File::Basename;
+  my $PREFIX = dirname(dirname(__FILE__));
+  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || "$PREFIX/share/autoconf";
   unshift @INC, $pkgdatadir;
 
   $buildauxdir = $ENV{'autom4te_buildauxdir'} || $pkgdatadir . '/build-aux';
@@ -117,9 +119,9 @@ Written by David J. MacKenzie and Akim Demaille.
 ";
 
 # Lib files.
-my $autoconf    = $ENV{'AUTOCONF'}    || '@bindir@/@autoconf-name@';
-my $autoheader  = $ENV{'AUTOHEADER'}  || '@bindir@/@autoheader-name@';
-my $autom4te    = $ENV{'AUTOM4TE'}    || '@bindir@/@autom4te-name@';
+my $autoconf    = $ENV{'AUTOCONF'}    || '@autoconf-name@';
+my $autoheader  = $ENV{'AUTOHEADER'}  || '@autoheader-name@';
+my $autom4te    = $ENV{'AUTOM4TE'}    || '@autom4te-name@';
 my $automake    = $ENV{'AUTOMAKE'}    || 'automake';
 my $aclocal     = $ENV{'ACLOCAL'}     || 'aclocal';
 my $libtoolize  = $ENV{'LIBTOOLIZE'}  || 'libtoolize';
diff --git a/bin/autoscan.in b/bin/autoscan.in
index b89fd1a..80968f1 100644
--- a/bin/autoscan.in
+++ b/bin/autoscan.in
@@ -30,7 +30,9 @@ use warnings FATAL => 'all';
 
 BEGIN
 {
-  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || '@pkgdatadir@';
+  use File::Basename;
+  my $PREFIX = dirname(dirname(__FILE__));
+  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || "$PREFIX/share/autoconf";
   unshift @INC, $pkgdatadir;
 
   # Override SHELL.  On DJGPP SHELL may not be set to a shell
@@ -95,10 +97,10 @@ my %needed_macros =
 my $log;
 
 # Autoconf and lib files.
-my $autom4te = $ENV{'AUTOM4TE'} || '@bindir@/@autom4te-name@';
+my $autom4te = $ENV{'AUTOM4TE'} || '@autom4te-name@';
 my $autoconf = "$autom4te --language=autoconf";
 my @prepend_include;
-my @include = ('@pkgdatadir@');
+my @include = ("$SHARE_DIR");
 
 # $help
 # -----
diff --git a/bin/autoupdate.in b/bin/autoupdate.in
index c86203a..1866dcb 100644
--- a/bin/autoupdate.in
+++ b/bin/autoupdate.in
@@ -29,9 +29,12 @@ use 5.006;
 use strict;
 use warnings FATAL => 'all';
 
+
 BEGIN
 {
-  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || '@pkgdatadir@';
+  use File::Basename;
+  my $PREFIX = dirname(dirname(__FILE__));
+  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || "$PREFIX/share/autoconf";
   unshift @INC, $pkgdatadir;
 
   # Override SHELL.  On DJGPP SHELL may not be set to a shell
@@ -50,12 +53,15 @@ use Autom4te::FileUtils;
 use Autom4te::General;
 use Autom4te::XFile;
 
+my $PREFIX = dirname(dirname(__FILE__));
+my $SHARE_DIR = "$PREFIX/share/autoconf";
+
 # Lib files.
-my $autom4te = $ENV{'AUTOM4TE'} || '@bindir@/@autom4te-name@';
+my $autom4te = $ENV{'AUTOM4TE'} || '@autom4te-name@';
 my $autoconf = "$autom4te --language=autoconf";
 # We need to find m4sugar.
 my @prepend_include;
-my @include = ('@pkgdatadir@');
+my @include = ("$SHARE_DIR");
 my $force = 0;
 # m4.
 my $m4 = $ENV{"M4"} || '@M4@';
diff --git a/bin/ifnames.in b/bin/ifnames.in
index b04947f..50ac4b4 100644
--- a/bin/ifnames.in
+++ b/bin/ifnames.in
@@ -35,7 +35,9 @@ use warnings FATAL => 'all';
 
 BEGIN
 {
-  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || '@pkgdatadir@';
+  use File::Basename;
+  my $PREFIX = dirname(dirname(__FILE__));
+  my $pkgdatadir = $ENV{'autom4te_perllibdir'} || "$PREFIX/share/autoconf";
   unshift @INC, $pkgdatadir;
 
   # Override SHELL.  On DJGPP SHELL may not be set to a shell
diff --git a/lib/autom4te.in b/lib/autom4te.in
index 9e86c9f..ce069a7 100644
--- a/lib/autom4te.in
+++ b/lib/autom4te.in
@@ -115,7 +115,6 @@ end-language: "Autoreconf-preselections"
 # This intermediate language is used by aclocal to build aclocal.m4.
 
 begin-language: "Autoconf-without-aclocal-m4"
-args: --prepend-include '@pkgdatadir@'
 args: --cache=autom4te.cache
 args: autoconf/autoconf.m4f
 args: acsite.m4?
@@ -142,7 +141,6 @@ end-language: "Autoconf"
 ## -------- ##
 
 begin-language: "Autotest"
-args: --prepend-include '@pkgdatadir@'
 args: autotest/autotest.m4f
 args: package.m4?
 args: local.at?
@@ -156,7 +154,6 @@ end-language: "Autotest"
 ## ---- ##
 
 begin-language: "M4sh"
-args: --prepend-include '@pkgdatadir@'
 args: m4sugar/m4sh.m4f
 args: --mode 777
 args: --language M4sugar
@@ -168,6 +165,5 @@ end-language: "M4sh"
 ## ------- ##
 
 begin-language: "M4sugar"
-args: --prepend-include '@pkgdatadir@'
 args: m4sugar/m4sugar.m4f
 end-language: "M4sugar"
diff --git a/man/autoconf.1 b/man/autoconf.1
index dcf6762..50d94bb 100644
--- a/man/autoconf.1
+++ b/man/autoconf.1
@@ -1,5 +1,5 @@
-.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.17.
-.TH AUTOCONF "1" "January 2021" "GNU Autoconf 2.71" "User Commands"
+.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
+.TH AUTOCONF "1" "February 2024" "GNU Autoconf 2.71" "User Commands"
 .SH NAME
 autoconf \- Generate configuration scripts
 .SH SYNOPSIS
diff --git a/man/autoheader.1 b/man/autoheader.1
index d90418b..52dcfdf 100644
--- a/man/autoheader.1
+++ b/man/autoheader.1
@@ -1,5 +1,5 @@
-.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.17.
-.TH AUTOHEADER "1" "January 2021" "GNU Autoconf 2.71" "User Commands"
+.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
+.TH AUTOHEADER "1" "February 2024" "GNU Autoconf 2.71" "User Commands"
 .SH NAME
 autoheader \- Create a template header for configure
 .SH SYNOPSIS
diff --git a/man/autom4te.1 b/man/autom4te.1
index c7902d2..25071e2 100644
--- a/man/autom4te.1
+++ b/man/autom4te.1
@@ -1,5 +1,5 @@
-.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.17.
-.TH AUTOM4TE "1" "January 2021" "GNU Autoconf 2.71" "User Commands"
+.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
+.TH AUTOM4TE "1" "February 2024" "GNU Autoconf 2.71" "User Commands"
 .SH NAME
 autom4te \- Generate files and scripts thanks to M4
 .SH SYNOPSIS
diff --git a/man/autoreconf.1 b/man/autoreconf.1
index a0970b1..f58102b 100644
--- a/man/autoreconf.1
+++ b/man/autoreconf.1
@@ -1,5 +1,5 @@
-.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.17.
-.TH AUTORECONF "1" "January 2021" "GNU Autoconf 2.71" "User Commands"
+.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
+.TH AUTORECONF "1" "February 2024" "GNU Autoconf 2.71" "User Commands"
 .SH NAME
 autoreconf \- Update generated configuration files
 .SH SYNOPSIS
diff --git a/man/autoscan.1 b/man/autoscan.1
index cae6400..26aa1a5 100644
--- a/man/autoscan.1
+++ b/man/autoscan.1
@@ -1,5 +1,5 @@
-.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.17.
-.TH AUTOSCAN "1" "January 2021" "GNU Autoconf 2.71" "User Commands"
+.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
+.TH AUTOSCAN "1" "February 2024" "GNU Autoconf 2.71" "User Commands"
 .SH NAME
 autoscan \- Generate a preliminary configure.ac
 .SH SYNOPSIS
diff --git a/man/autoupdate.1 b/man/autoupdate.1
index 0b18ca5..4b36931 100644
--- a/man/autoupdate.1
+++ b/man/autoupdate.1
@@ -1,5 +1,5 @@
-.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.17.
-.TH AUTOUPDATE "1" "January 2021" "GNU Autoconf 2.71" "User Commands"
+.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
+.TH AUTOUPDATE "1" "February 2024" "GNU Autoconf 2.71" "User Commands"
 .SH NAME
 autoupdate \- Update a configure.ac to a newer Autoconf
 .SH SYNOPSIS
diff --git a/man/ifnames.1 b/man/ifnames.1
index 0bb65d8..ca768eb 100644
--- a/man/ifnames.1
+++ b/man/ifnames.1
@@ -1,5 +1,5 @@
-.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.17.
-.TH IFNAMES "1" "January 2021" "GNU Autoconf 2.71" "User Commands"
+.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
+.TH IFNAMES "1" "February 2024" "GNU Autoconf 2.71" "User Commands"
 .SH NAME
 ifnames \- Extract CPP conditionals from a set of files
 .SH SYNOPSIS
-- 
2.44.0.278.ge034bb2e1d-goog

